src/DataGenerator.cpp

```
@@ -0,0 +1,499 @@
+    node_criteria = doc.child("dataset").child("criteria");
+  }
+
+  std::stringstream strValue;
+  int nb_crit;
+  strValue << node_criteria.child_value();
+  strValue >> nb_crit;
+  return nb_crit;
+}
+
+int DataGenerator::getNumberOfCategories(std::string fileName) {
+  pugi::xml_document doc;
+  std::string path = "../data/" + fileName;
+  if (!doc.load_file(path.c_str()))
+    throw std::invalid_argument("Cannot open xml file, please check path");
+
+  pugi::xml_node node_cat;
+  if (DataGenerator::getXmlFileType(fileName) == "model") {
+    // if we have a model xml
+    node_cat = doc.child("model").child("categories");
+  } else {
+    // if we have a dataset xml
+    node_cat = doc.child("dataset").child("categories");
+  }
+
+  std::stringstream strValue;
+  int nb_categories;
+  strValue << node_cat.child_value();
+  strValue >> nb_categories;
+  return nb_categories;
+}
+
+float DataGenerator::getThresholdValue(std::string fileName) {
+  pugi::xml_document doc;
+  std::string path = "../data/" + fileName;
+  if (!doc.load_file(path.c_str()))
+    throw std::invalid_argument("Cannot open xml file, please check path");
+
+  if (DataGenerator::getXmlFileType(fileName) == "model") {
+    // if we have a model xml
+    pugi::xml_node lambda_node = doc.child("model").child("lambda");
+    float lambda = atof(lambda_node.child_value());
+    return lambda;
+  } else {
+    // if we have a dataset xml
+    throw std::invalid_argument("Cannot find any threshold in xml file, "
+                                "most likely have a xml model file");
+  }
+}
+
+int DataGenerator::getNumberOfAlternatives(std::string fileName) {
+  pugi::xml_document doc;
+  std::string path = "../data/" + fileName;
+  if (!doc.load_file(path.c_str()))
+    throw std::invalid_argument("Cannot open xml file, please check path");
+
+  if (DataGenerator::getXmlFileType(fileName) == "model") {
+    // if we have a model xml
+    throw std::invalid_argument(
```

same as above

---

Exception in getXmlFileType should throw an exception so we should be fine